internal/pkg/service/common/etcdop/op/atomic.go

```diff
@@ -94,6 +83,17 @@ func (v *AtomicOp[R]) WritePhaseOps() (out []HighLevelFactory) {
 	return out
 }
 
+// Core returns a common interface of the atomic operation, without result type specific methods.
+// It is useful when you need to use some helper/hook to modify atomic operations with different result types.
+func (v *AtomicOp[R]) Core() *AtomicOpCore {
+	return v.AtomicOpCore
+}
+
+func (v *AtomicOp[R]) AddFrom(ops ...AtomicOpInterface) *AtomicOp[R] {
+	v.AtomicOpCore.AddFrom(ops...)
+	return v
+}
+
```

Comment:
`AtomicOpCore` methods return `*AtomicOpCore`, and `AtomicOp[R]` returns `*AtomicOp[R]`.

So methods of both structs can be chained: `o.RequireLock(...).ReadOrErr(...)...`.